src/Analyzer/Passes/LogicalExpressionOptimizer.cpp

```diff
@@ -0,0 +1,224 @@
+            const auto & rhs = equals_arguments[1];
+
+            const auto add_equals_function_if_not_present = [&](const auto & expression_node, const ConstantNode * constant)
+            {
+                auto & constant_set = node_to_constants[expression_node];
+                if (!constant_set.contains(constant))
+                {
+                    constant_set.insert(constant);
+                    node_to_equals_functions[expression_node].push_back(argument);
+                }
+            };
+
+            if (const auto * lhs_literal = lhs->as<ConstantNode>())
+                add_equals_function_if_not_present(rhs, lhs_literal);
+            else if (const auto * rhs_literal = rhs->as<ConstantNode>())
+                add_equals_function_if_not_present(lhs, rhs_literal);
+            else
+                or_operands.push_back(argument);
+        }
+
+        auto in_function_resolver = FunctionFactory::instance().get("in", getContext());
+
+        for (auto & [expression, equals_functions] : node_to_equals_functions)
+        {
+            const auto & settings = getSettings();
+            if (equals_functions.size() < settings.optimize_min_equality_disjunction_chain_length && !expression.node->getResultType()->lowCardinality())
+            {
+                std::move(equals_functions.begin(), equals_functions.end(), std::back_inserter(or_operands));
+                continue;
+            }
+
+            Tuple args;
+            args.reserve(equals_functions.size());
+            /// first we create tuple from RHS of equals functions
+            for (const auto & equals : equals_functions)
+            {
+                const auto * equals_function = equals->as<FunctionNode>();
+                assert(equals_function && equals_function->getFunctionName() == "equals");
+
+                const auto & equals_arguments = equals_function->getArguments().getNodes();
+                if (const auto * rhs_literal = equals_arguments[1]->as<ConstantNode>())
+                {
+                    args.push_back(rhs_literal->getValue());
+                }
+                else
+                {
+                    const auto * lhs_literal = equals_arguments[0]->as<ConstantNode>();
+                    assert(lhs_literal);
+                    args.push_back(lhs_literal->getValue());
+                }
+            }
+
+            auto rhs_node = std::make_shared<ConstantNode>(std::move(args));
+
+            auto in_function = std::make_shared<FunctionNode>("in");
+
+            QueryTreeNodes in_arguments;
+            in_arguments.reserve(2);
+            in_arguments.push_back(expression.node);
+            in_arguments.push_back(std::move(rhs_node));
+
+            in_function->getArguments().getNodes() = std::move(in_arguments);
+            in_function->resolveAsFunction(in_function_resolver);
+
+            or_operands.push_back(std::move(in_function));
+        }
+
+        if (or_operands.size() == 1)
```

Comment:
The same point as for `and`

Reply:
Well, I agree that the result type for `and` must be okay, but I'm afraid that for `or` we can suddenly change it. So I'd prefer to add a check here.